### PR TITLE
trilium-next: init at 0.90.4

### LINKED ISSUE #339237: Package request: TriliumNext-server and TriliumNext-desktop
TriliumNext Notes is an open-source, cross-platform hierarchical note taking application with focus on building large personal knowledge bases.

**Metadata**
* source URL: https://github.com/TriliumNext/Notes
* license: AGPL-3,0
* platforms: linux, windows, mac

TriliumNext is the successor of Trilium which is in maintenance mode now (https://github.com/zadam/trilium/issues/4620). I've already posted a request to update for the original Trilium package (https://github.com/NixOS/nixpkgs/issues/339039) but it's seems to be a possible solution not to silently update the package to a fork but providing a separate package for TriliumNext.

The fork seems to be in active and well maintained. I already use TriliumNext on my arch-based client and would like to migrate my nix-based server instance to TriliumNext.

### PR DESCRIPTION
**Pull request by eliandoran**
This PR adds [TriliumNext Notes](https://github.com/TriliumNext/Notes) which is a fork of Trilium Notes, which has entered maintenance mode.

Basically I've taken the original package description for Trilium and adapted it for the fork. I've also changed the server module to take in a `package` attribute.

Note: I am also the upstream maintainer of TriliumNext Notes.

Closes #339237

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/trilium.nix [hunk 1]
```diff
@@ -117,7 +119,7 @@ in
       wantedBy = [ "multi-user.target" ];
       environment.TRILIUM_DATA_DIR = cfg.dataDir;
       serviceConfig = {
-        ExecStart = "${pkgs.trilium-server}/bin/trilium-server";
+        ExecStart = "${cfg.package}/bin/trilium-server";
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/trilium.nix:122]:
```suggestion
        ExecStart = lib.getExe cfg.package;
```

File: pkgs/applications/office/trilium-next/default.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  trilium-next-desktop = callPackage ./desktop.nix {
+    metaCommon = metaCommon;
+    version = version;
+  };
+  trilium-next-server = callPackage ./server.nix {
+    metaCommon = metaCommon;
+    version = version;
+  };
+
+}
```
Comment (SuperSandro2000) [inline pkgs/applications/office/trilium-next/default.nix:25]:
```suggestion
{ lib, callPackage }:

let
  meta = with lib; {
    description = "Hierarchical note taking application with focus on building large personal knowledge bases";
    homepage = "https://github.com/TriliumNext/Notes";
    license = licenses.agpl3Plus;
    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
    platforms = [ "x86_64-linux" ];
    maintainers = with maintainers; [ eliandoran ];
  };
  version = "0.90.4";
in
{
  trilium-next-desktop = callPackage ./desktop.nix {
    inherit meta version;
  };

  trilium-next-server = callPackage ./server.nix {
    inherit meta version;
  };
}
```

File: pkgs/applications/office/trilium-next/desktop.nix [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      "x86_64-darwin"
+    ];
+  };
+
+  linux = stdenv.mkDerivation rec {
+    inherit pname version meta;
+
+    src = fetchurl linuxSource;
+
+    # TODO: migrate off autoPatchelfHook and use nixpkgs' electron
```
Comment (SuperSandro2000) [inline pkgs/applications/office/trilium-next/desktop.nix:43]: so, what's blocking that?

File: pkgs/applications/office/trilium-next/server.nix [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    runHook preInstall
+    mkdir -p $out/bin
+    mkdir -p $out/share/trilium-server
+
+    cp -r ./* $out/share/trilium-server
+    runHook postInstall
+  '';
+
+  postFixup = ''
+    cat > $out/bin/trilium-server <<EOF
```
Comment (SuperSandro2000) [inline pkgs/applications/office/trilium-next/server.nix:46]: Please use makeWrapper

File: pkgs/applications/office/trilium-next/desktop.nix [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    nativeBuildInputs = [ unzip ];
+
+    installPhase = ''
+      mkdir -p $out/Applications
+      cp -r *.app $out/Applications
+    '';
+  };
+
+in
+if stdenv.isDarwin then darwin else linux
```
Comment (FliegendeWurst) [inline pkgs/applications/office/trilium-next/desktop.nix:114]:
```suggestion
if stdenv.hostPlatform.isDarwin then darwin else linux
```

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -13066,6 +13066,11 @@ with pkgs;
     trilium-server
     ;
 
+  inherit (callPackage ../applications/office/trilium-next {})
+    trilium-next-desktop
+    trilium-next-server
+    ;
```
Comment (FliegendeWurst) [inline pkgs/top-level/all-packages.nix:13072]: should be in pkgs/by-name

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Comment by fabian-mv** [comment]
sadly this wont be coming out in 24.11, since [branch out was 4 days ago](https://github.com/NixOS/nixpkgs/issues/339153). this is very bad, as trilium next is now more important as sync became incompatible.

this means that trilium next won't be in nixpkgs until mid-2025 (since releases are very 6months). pretty sad if you ask me

@eliandoran i can help if you need

**Comment by FliegendeWurst** [comment]
New packages can always be backported. No problem there.

**Comment by FliegendeWurst** [comment]
I tried to package using nixpkgs' electron instead: https://github.com/FliegendeWurst/nixpkgs/blob/trilium-next/pkgs/applications/office/trilium-next/desktop.nix

... but for some reason, locale data is missing:

Screenshot_20241118_092017

@eliandoran do you know how this could be fixed? Do I need to add some more flags to my electron invocation?

**Comment by FliegendeWurst** [comment]
I fixed the above issue and created #356930

**Comment by eliandoran** [comment]
Closing in favor of https://github.com/NixOS/nixpkgs/pull/356930.

**Pull request closed without merging**